### ISSUE TITLE
SRS HLS 切片支持AES128 加密么？

### ISSUE THREAD

**Issue by harlanc** [ossrs/srs#1093]
nginx-rtmp-module 是支持的，貌似在SRS里面没有发现类似的选项？？

**Comment by winlinvip** [comment]
SRS3 feature

**Comment by winlinvip** [comment]
You can update the key, for example: 

```#EXTM3U
#EXT-X-VERSION:3
#EXT-X-ALLOW-CACHE:YES
#EXT-X-MEDIA-SEQUENCE:13
#EXT-X-TARGETDURATION:15
#EXTINF:10.147, no desc
livestream-13.ts
#EXTINF:10.007, no desc
livestream-14.ts
#EXT-X-KEY:METHOD=AES-128,URI="livestream-15.key",IV=0x175D2375FEBAE7ED8E92A7E6A64F5113
#EXTINF:10.216, no desc
livestream-15.ts
#EXTINF:10.170, no desc
livestream-16.ts
#EXTINF:11.378, no desc
livestream-17.ts
#EXTINF:9.668, no desc
livestream-18.ts
```

So it will directly request from 15:

![image](https://user-images.githubusercontent.com/2777660/44617117-e7599e80-a88f-11e8-816b-717f207dcc3c.png)


`TRANS_BY_GPT3`

**Comment by winlinvip** [comment]
Merged #1109.

**Comment by winlinvip** [comment]
To Use HLS enctyption:

```
listen              1935;
max_connections     1000;
daemon              off;
srs_log_tank        console;
http_server {
    enabled         on;
    listen          8080;
    dir             ./objs/nginx/html;
}
vhost __defaultVhost__ {
    hls {
        enabled         on;
        hls_fragment    10;
        hls_window      60;
        hls_path        ./objs/nginx/html;
        hls_m3u8_file   [app]/[stream].m3u8;
        hls_ts_file     [app]/[stream]-[seq].ts;
        hls_keys        on;
        hls_fragments_per_key 5;
        hls_key_file     [app]/[stream]-[seq].key;
        hls_key_file_path    ./objs/nginx/html;
        hls_key_url       http://localhost:8080/;
    }
    ingest livestream {
        enabled      on;
        input {
            type    file;
            url     ./doc/source.200kbps.768x320.flv;
        }
        ffmpeg      ./objs/ffmpeg/bin/ffmpeg;
        engine {
            enabled          off;
            output          rtmp://127.0.0.1:[port]/live?vhost=[vhost]/livestream;
        }
    }
}
```

Play HLS with encryption in Safari: http://localhost:8080/live/livestream.m3u8

**Comment by bml1g12** [comment]
I'm using the following config file:
```
listen              1935;
max_connections     1000;
daemon              off;
srs_log_tank        console;
http_server {
    enabled         on;
    listen          8080;
    dir             ./objs/nginx/html;
}
vhost __defaultVhost__ {
    hls {
        enabled         on;
        hls_fragment    10;
        hls_window      60;
        hls_path        ./objs/nginx/html;
        hls_m3u8_file   [app]/[stream].m3u8;
        hls_ts_file     [app]/[stream]-[seq].ts;
        hls_keys        on;
        hls_fragments_per_key 5;
        hls_key_file     [app]/[stream]-[seq].key;
        hls_key_file_path    ./objs/nginx/html;
    }
}
```
And I can access http://192.168.100.66/live/livestream.m3u8 and http://192.168.100.66/live/livestream-50.key successfully, so NGINX is working. When I load the stream in vlc ( http://0.0.0.0/live/livestream.m3u8 or http://192.168.100.66/live/livestream.m3u8 ) however the corresponding .m3u8 file seems to have the wrong path: 
```#EXTM3U
#EXT-X-VERSION:3
#EXT-X-ALLOW-CACHE:YES
#EXT-X-MEDIA-SEQUENCE:0
#EXT-X-TARGETDURATION:17
#EXT-X-DISCONTINUITY
#EXT-X-KEY:METHOD=AES-128,URI="./objs/nginx/htmllive/livestream-0.key",IV=0x655D870530FFA94DFE55CD152092C42A
#EXTINF:16.625, no desc
livestream-0.ts
```
So when I load it in VLC it looks for the non-existent path `http://192.168.100.66:80/live/objs/nginx/htmllive/livestream-50.key` for the key (it looks in the correct place for the .m3u8 file). 

I have tried changing `hls_key_file_path` (e.g. to `./`) but it seems to have no effect on the URI field of the `.m3u8` file. 

How do I configure the path auto-generated in the `.m3u8` URI field?

EDIT RESOLVED: by adding `hls_key_url http://0.0.0.0/;`

`TRANS_BY_GPT3`
